### PR TITLE
Added exercises for Typing React State

### PR DESCRIPTION
**Pull request by avinavdixit25**

### CODE REVIEW

**Comment by chasenlehara** [comment]
Adding the links here so this is easier to review:


**Exercise 1**

- [Problem](https://codesandbox.io/s/github/bitovi/trainings/tree/typing-react-state/typing-react-state/Exercise1/problem?file=/App.tsx)
- [Solution](https://codesandbox.io/s/github/bitovi/trainings/tree/typing-react-state/typing-react-state/Exercise1/solution?file=/App.tsx)

**Exercise 2**

- [Problem](https://codesandbox.io/s/github/bitovi/trainings/tree/typing-react-state/typing-react-state/Exercise2/problem?file=/App.tsx)
- [Solution](https://codesandbox.io/s/github/bitovi/trainings/tree/typing-react-state/typing-react-state/Exercise2/solution?file=/App.tsx)

**Exercise 3**

- [Problem](https://codesandbox.io/s/github/bitovi/trainings/tree/typing-react-state/typing-react-state/Exercise3/problem?file=/App.tsx)
- [Solution](https://codesandbox.io/s/github/bitovi/trainings/tree/typing-react-state/typing-react-state/Exercise3/solution?file=/App.tsx)

File: typing-react-state/Exercise1/problem/.codesandbox/template.json [hunk 1]
```diff
@@ -0,0 +1,10 @@
+{
+  "title": "Exercise 1",
+  "description": "First exercise for Typing React State",
+  "iconUrl": "https://www.bitovi.com/favicon.ico",
+  "tags": [
+    "Typescript",
+    "React"
```
Comment (chasenlehara) [inline typing-react-state/Exercise1/problem/.codesandbox/template.json:7]: These can be `["react", "typescript"]`

File: typing-react-state/Exercise1/problem/package.json [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+  "keywords": [
+    "typescript",
+    "react",
+    "starter"
+  ],
+  "main": "index.tsx",
+  "dependencies": {
+    "react": "17.0.2",
+    "react-dom": "17.0.2",
+    "react-scripts": "4.0.0"
```
Comment (chasenlehara) [inline typing-react-state/Exercise1/problem/package.json:14]: Instead of “pinning” the versions here, we should have the normal `^` semver range and commit the `package-lock.json` file.

File: typing-react-state/Exercise1/solution/App.tsx [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+      owner: "Michael P.",
+      departments: ["React", "Angular", "DevOps"]
+    });
+
+    setEmployees(["Alice", "Bob", "Cathy"]);
+  }, []);
+
+  return (
+    <>
+      <section>Company Name: {companyName}</section>
```
Comment (chasenlehara) [inline typing-react-state/Exercise1/solution/App.tsx:45]: These should be `p` elements.

File: typing-react-state/Exercise1/solution/App.tsx [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+      <section>
+        Company Description:{" "}
+        {companyDescription ? companyDescription : "No description"}
+        <div>
+          <button onClick={() => setCompanyDescription(null)}>
+            Reset Company Description
+          </button>
+        </div>
```
Comment (chasenlehara) [inline typing-react-state/Exercise1/solution/App.tsx:53]: I think this `div` can be removed.

File: typing-react-state/Exercise1/solution/App.tsx [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+      </section>
+      <section>
+        <div> Size: {companyDetails.size} </div>
+
+        <div>
+          <div> Owner: {companyDetails.owner || " Not Known"} </div>
```
Comment (chasenlehara) [inline typing-react-state/Exercise1/solution/App.tsx:59]: These should be `p` elements.

File: typing-react-state/Exercise1/solution/App.tsx [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+        </div>
+
+        <div>
+          Departments:
+          {companyDetails.departments?.map((department) => (
+            <div key={department}>{department}</div>
+          ))}
+          {/* Property 'departments' does not exist on type '{}'. */}
```
Comment (chasenlehara) [inline typing-react-state/Exercise1/solution/App.tsx:67]: Is this supposed to be here?

File: typing-react-state/Exercise1/solution/App.tsx [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+            }
+          >
+            Reset Company Details
+          </button>
+        </div>
+      </section>
+      <section>
+        Employee List:
+        {employees ? (
+          employees?.map((employee) => <div key={employee}>{employee}</div>)
```
Comment (chasenlehara) [inline typing-react-state/Exercise1/solution/App.tsx:86]: These should be `li` elements wrapped in an `ol` or `ul`.

File: typing-react-state/Exercise1/solution/App.tsx [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+        ) : (
+          <div>No employees</div>
```
Comment (chasenlehara) [inline typing-react-state/Exercise1/solution/App.tsx:88]: This should be `p` element.

File: typing-react-state/Exercise1/solution/App.tsx [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+        )}
+        <div>
+          <button onClick={() => setEmployees(undefined)}>
+            Reset Employee List
+          </button>
+        </div>
```
Comment (chasenlehara) [inline typing-react-state/Exercise1/solution/App.tsx:94]: Does this `div` do anything?

File: typing-react-state/Exercise2/problem/App.tsx [hunk 10]
```diff
@@ -0,0 +1,82 @@
+// Correct the Action type so that TypeScript can
+// catch errors in the second and third dispatch calls.
+// If the Action type is correctly written -
+//    - you will need to make changes to the second
+//      dispatch call in order to correct the error
+//      that TypeScript shows. Clicking the "Reset" button
+//      should reset "name" to an empty string.
+//    - TypeScript will complain about the “type” property
+//      in the third dispatch call. This dispatch will
+//      have to be deleted entirely.
```
Comment (chasenlehara) [inline typing-react-state/Exercise2/problem/App.tsx:10]: For all of these prompts, we may want to number them so the “minimum” we want them to do is more clear.

Example: https://github.com/bitovi/trainings/blob/299ff9d94a05206698d11061be9bc98e20c8f137/intro-to-accessibility-first/buttons-and-links/problem/index.html#L4-L9

File: typing-react-state/Exercise2/solution/App.tsx [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+  function appReducer(state: State, action: Action): State {
+    switch (action.type) {
+      case "updateName":
+        return { ...state, name: action.payload.name };
+      case "reset":
+        return { ...state, name: "" };
+      default:
+        return state;
+    }
+  }
```
Comment (chasenlehara) [inline typing-react-state/Exercise2/solution/App.tsx:32]: We should move this out of `App`’s scope, right?

File: typing-react-state/Exercise3/solution/App.tsx [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+  const lastNameRef = useRef({
+    lastName: ""
+  });
+
+  const handleFirstNameChange = (e: React.ChangeEvent<HTMLInputElement>) => {
+    const firstName = e.target.value;
+    firstNameRef.current = firstName
+      ? firstName.charAt(0).toUpperCase() + firstName.slice(1)
+      : null;
+    // Cannot assign to 'current' because it is a read-only property.
```
Comment (chasenlehara) [inline typing-react-state/Exercise3/solution/App.tsx:21]: Ah, I think I understand these now. I think it’s safe to remove these from both the problem and solution, since CodeSandbox will show these errors.

File: typing-react-state/Exercise3/solution/App.tsx [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+    // Property 'lastName' does not exist on type '{}'.
+    setFullName(fullName);
+  };
+
+  return (
+    <>
+      <div>
+        First Name:
+        <input onChange={handleFirstNameChange} />
+      </div>
```
Comment (chasenlehara) [inline typing-react-state/Exercise3/solution/App.tsx:46]: These inputs should be wrapped in `label` elements.

**Review by chasenlehara — CHANGES_REQUESTED** [review]
It would be good to have all the necessary setup to run these locally so we can easily develop them without relying on CodeSandbox.

Please feel free to merge in as-is, or make changes in this branch and then merge it in.

File: typing-react-state/Exercise1/problem/App.tsx [hunk 14]
```diff
@@ -0,0 +1,88 @@
+// Add/Correct the types and get rid of TypeScript errors
+// Things to take care of -
+//  - companyDescription should be allowed to be set to null
+//  - The "Reset Company Details" button has a setCompanyDetails call. Something here
+//    will have to be changed after you have added the correct type to the
+//    useState hook for companyDetails.
+
+import "./styles.css";
+import React, { useState, useEffect } from "react";
```
Comment (CMLabay) [inline typing-react-state/Exercise1/problem/App.tsx:9]: remove React import

File: typing-react-state/Exercise1/solution/App.tsx [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+// Things to take care of -
+//  - companyDescription should be allowed to be set to null
+//  - The "Reset Company Details" button has a setCompanyDetails call. Something here
+//    will have to be changed after you have added the correct type to the
+//    useState hook for companyDetails.
+
+import "./styles.css";
+import React, { useState, useEffect } from "react";
+
+interface CompanyDetailProps {
```
Comment (CMLabay) [inline typing-react-state/Exercise1/solution/App.tsx:11]: CompanyDetails

File: typing-react-state/Exercise1/solution/App.tsx [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+  departments: string[];
+};
+
+const App = () => {
+  const [companyName, setCompanyName] = useState("");
+  const [companyDescription, setCompanyDescription] = useState<string | null>(
+    null
+  );
+  const [companyDetails, setCompanyDetails] = useState<CompanyDetailProps>({
+    size: 0,
```
Comment (CMLabay) [inline typing-react-state/Exercise1/solution/App.tsx:23]: make a variable for initial details

File: typing-react-state/Exercise2/problem/App.tsx [hunk 17]
```diff
@@ -0,0 +1,82 @@
+// Correct the Action type so that TypeScript can
+// catch errors in the second and third dispatch calls.
+// If the Action type is correctly written -
+//    - you will need to make changes to the second
+//      dispatch call in order to correct the error
+//      that TypeScript shows. Clicking the "Reset" button
+//      should reset "name" to an empty string.
+//    - TypeScript will complain about the “type” property
```
Comment (CMLabay) [inline typing-react-state/Exercise2/problem/App.tsx:8]: reword this to be less specific

File: typing-react-state/Exercise3/problem/App.tsx [hunk 18]
```diff
@@ -0,0 +1,57 @@
+// Make changes to the useRefs in order to correct TS errors
+// Take note of the state's structure while adding types.
+// Bonus question - What happens if we use useState
+// instead of useRef?
```
Comment (CMLabay) [inline typing-react-state/Exercise3/problem/App.tsx:4]: what if similar to slide ex useref to give focus to first input that has an error?

**Comment by chasenlehara** [comment]
I’m going to merge this in as-is. @CMLabay and I will work through our review comments.
